src/main/java/net/minecraftforge/network/HandshakeHandler.java

```diff
@@ -329,6 +331,78 @@ void handleConfigSync(final HandshakeMessages.S2CConfigData msg, final Supplier<
         contextSupplier.get().setPacketHandled(true);
         NetworkConstants.handshakeChannel.reply(new HandshakeMessages.C2SAcknowledge(), contextSupplier.get());
     }
+
+    void handleClientReset(HandshakeMessages.S2CReset msg, final Supplier<NetworkEvent.Context> contextSupplier) {
+        NetworkEvent.Context context = contextSupplier.get();
+        Connection connection = context.getNetworkManager();
+
+        if (context.getDirection() != NetworkDirection.LOGIN_TO_CLIENT && context.getDirection() != NetworkDirection.PLAY_TO_CLIENT) {
+            connection.disconnect(Component.literal("Illegal packet received, terminating connection"));
+            throw new IllegalStateException("Invalid packet received, aborting connection");
+        }
+
+        LOGGER.debug(FMLHSMARKER, "Received reset from server");
+
+        ServerData serverData = Minecraft.getInstance().getCurrentServer();
+        Screen screen = Minecraft.getInstance().screen;
+
+        if (!handleClear(context)) {
+            return;
+        }
+
+        NetworkHooks.registerClientLoginChannel(connection);
+        connection.setProtocol(ConnectionProtocol.LOGIN);
+        connection.setListener(new ClientHandshakePacketListenerImpl(
+                connection, Minecraft.getInstance(), serverData, screen, true, null, statusMessage -> {}
+        ));
+
+        context.setPacketHandled(true);
+        NetworkConstants.handshakeChannel.reply(
+                new HandshakeMessages.C2SAcknowledge(),
+                new NetworkEvent.Context(connection, NetworkDirection.LOGIN_TO_CLIENT, 98)
+        );
+
+        LOGGER.debug(FMLHSMARKER, "Reset complete");
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    boolean handleClear(NetworkEvent.Context context) {
+        CompletableFuture<Void> future = context.enqueueWork(() -> {
+            LOGGER.debug(FMLHSMARKER, "Clearing");
+
+            // Clear
+            if (Minecraft.getInstance().level == null) {
+                // Ensure the GameData is reverted in case the client is reset during the handshake.
+                GameData.revertToFrozen();
+            }
+
+            // Clear
+            Minecraft.getInstance().clearLevel(new GenericDirtMessageScreen(Component.translatable("connect.negotiating")));
+            try {
+                context.getNetworkManager().channel().pipeline().remove("forge:forge_fixes");
+            } catch (NoSuchElementException ignored) {
+            }
+            try {
+                context.getNetworkManager().channel().pipeline().remove("forge:vanilla_filter");
+            } catch (NoSuchElementException ignored) {
+            }
+            // Restore
+//            Minecraft.getInstance().setCurrentServer(serverData);
+        });
+
+        LOGGER.debug(FMLHSMARKER, "Waiting for clear to complete");
+
+        try {
+            future.get();
```

Comment:
Why use a CompletableFutureonly to immediately call .get()?

Reply:
So in all honest, and I think I have said this somewhere before, it's not actually my code I asked the original author if they were okay with me taking it forward (as they did not want to) and they said yes. So I'm not 100% certain on why it's using a Future here, but I suspect it's to make sure that the code executes on the render/main (client) thread? And then the .get() call would hold the thread until it's complete?

Reply:
I missed that the future was coming out of `enqueueWork` - that's expected (and correct).  It does force the main thread to run the block.

What is abnormal is blocking the network thread on completion of the future, but I guess in this case it has to, as the state of the connection is dependent of the outcome of the main-thread related work.